locallib.php

```diff
@@ -621,6 +624,259 @@ private function process_action_manual_allocation() {
+        }
+        foreach ($choicessortedwithgroupscount as &$choiceswithcertaingroupcount) {
+            usort($choiceswithcertaingroupcount, function($a, $b) use ($placesleft) {
+                // Choices with the same amount of groups are sorted according the count of left places: fewer places first.
+                return $placesleft[$a] - $placesleft[$b];
+            });
+        }
+        for ($i = 1; $i <= max(array_keys($choicessortedwithgroupscount)); $i++) {
+            if (empty($choicessortedwithgroupscount[$i])) {
+                continue;
+            }
+            $choicessorted = array_merge($choicessorted, $choicessortedwithgroupscount[$i]);
+        }
+        if (!empty($choicessortedwithgroupscount[0])) {
+            $choicessorted = array_merge($choicessorted, $choicessortedwithgroupscount[0]);
+        }
+
+        // This is kind of a dilemma. We want the choices to be filled up beginning at the one with the least places left to fill it
+        // up as quickly as possible.
+        // However, in case of group restrictions this will lead to problems as we are assigning users which have to be assigned to
+        // specific choices (because all others cannot be assigned to it). So in the end these choices will not be available when
+        // we arrive at the choice with the group restrictions.
+        // Therefore, we are first filling up choices with group restrictions first (beginning at choices with fewer groups). Only
+        // in case we have the same amount of groups for two choices or we have no group restrictions at all we pick choices with
+        // fewer places left first (see foreach loop with usort a few lines above).
+
+        return !empty($choicessorted) ? array_shift($choicessorted) : -1;
+    }
+
+    /**
+     * Try to distribute all currently unallocated users.
+     *
+     * @param string $distributionalgorithm the distributionalgorithm which should be used, you can choose between
+     *  ACTION_DISTRIBUTE_UNALLOCATED_EQUALLY and ACTION_DISTRIBUTE_UNALLOCATED_FILL
+     * @return void
+     * @throws dml_exception
+     */
+    public function distribute_users_without_choice(string $distributionalgorithm): void {
+        // This could need some extra memory, especially because we are caching some data structures in memory while
+        // running the algorithm.
+        raise_memory_limit(MEMORY_EXTRA);
+        core_php_time_limit::raise();
+
+        // This will retrieve a list of users we are trying to assign to choices, sorted from the least group membership count to
+        // more group memberships. Users without group memberships will be at the end of the array.
+        $possibleusers = $this->get_undistributed_users();
+
+        $transaction = $this->db->start_delegated_transaction();
+
+        $usertoassign = array_shift($possibleusers);
+        // As long as we have a user to assign, we try to assign him.
```

Comment:
The allocation with the 500 students worked however it still took several minutes. I noticed that there were more students than available choices in my course. In this scenario, all students were iterated even when there wasn't any open space at a choice available. This might be a scenario which does not occur regularly but should be considered to avoid unnecessary calculations. I would suggest adding a check how many places are left in general so that the algorithm can directly stop when all spots are filled.

Reply:
I had a look at the code and I think I know what you are talking about. I hope I got you right and I added a commit to address this. Thank you very much for this idea.

Also thank you very much doing a test with 500 students. 9 minutes is indeed a long time. It's quite unlikely that 500 students did not vote at all and all of them need to be distributed I'd think, but well...maybe there are teachers/instructors who just use this tool to distribute the students SOMEHOW without even letting them rate first. So, that definitely makes me think if it would be a good idea to run the algorithm as ad hoc task. This of course is not ideal for usability, because the user would press a button and then have to guess when the algorithm has finished, but it's probably better to hit the refresh button a few times than running into a timeout of the reverse proxy or something. Thoughts on this?

Reply:
Thank you! I have also already thought about running this as an adhoc task. In my opinion, this consideration also applies to the normal allocation algorithm which if not specified differently, is still run without cron. Hence, to be consistent I would leave this as it is and maybe create a new issue for moving everything which could take longer into a cron task.